### PR TITLE
search: paginated repo resolution

### PR DESCRIPTION
**Pull request by tsenart**

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search.go [hunk 1]
```diff
@@ -279,6 +279,10 @@ func (r *searchResolver) resolveRepositories(ctx context.Context, options search
 		tr.Finish()
 	}()
 
+	// TODO(tsenart)
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search.go:282]: @rvantonder: Any thoughts on this TODO? I know there is complicated cache invalidation logic, so I want to defer this to later.

File: cmd/frontend/graphqlbackend/search_alert.go [hunk 2]
```diff
@@ -299,6 +299,10 @@ func (r *searchResolver) errorForOverRepoLimit(ctx context.Context) *errOverRepo
 	}
 
 	repoOptions := r.toRepoOptions(r.Query, resolveRepositoriesOpts{})
+
+	// TODO(tsenart): Figure out what to do with this call of resolveRepositories.
+	//  Do we really need to potentially resolve ALL repos in the database to accomplish the goals of
+	//  this method? Or would limiting to a single page of say 500 repositories max do the trick?
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_alert.go:305]: @rvantonder: Question for you here :)

File: cmd/frontend/graphqlbackend/search_results.go [hunk 3]
```diff
@@ -1668,9 +1669,10 @@ func (r *searchResolver) doResults(ctx context.Context, args *search.TextParamet
 		wg.Add(1)
 		goroutine.Go(func() {
 			defer wg.Done()
+			// TODO(tsenart): Migrate this search type to use searchrepos.Pager instead of the already resolved repositories,
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_results.go:1672]: @rvantonder: Leaving this one untouched in this PR. I assume you'll follow up soon on moving this to a Job interface?

File: cmd/frontend/graphqlbackend/search_results.go [hunk 4]
```diff
@@ -1682,6 +1684,7 @@ func (r *searchResolver) doResults(ctx context.Context, args *search.TextParamet
 
 				notSearcherOnly := !searcherOnly
 
+				// TODO(tsenart): Migrate this search type to use searchrepos.Pager instead of the already resolved repositories.
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_results.go:1687]: @rvantonder: Same here. Leaving this one untouched in this PR.

Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_results.go:1687]: @rvantonder: Same here. Leaving this one untouched in this PR.

File: cmd/frontend/graphqlbackend/search_suggestions.go [hunk 5]
```diff
@@ -244,6 +249,8 @@ func (r *searchResolver) showRepoSuggestions(ctx context.Context) ([]SearchSugge
 				limit:                    maxSearchSuggestions,
 			})
 
+		// TODO(tsenart): Figure out what to do with this instance of resolveRepositories.
+		//  I think we're getting rid of GraphQL suggestions code, so this might be a non-issue.
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_suggestions.go:253]: @novoselrok: Care to comment?

File: internal/search/commit/commit_new.go [hunk 6]
```diff
@@ -29,53 +31,71 @@ type CommitSearch struct {
...
+		if err := CheckSearchLimits(j.HasTimeFilter, len(page.RepoRevs), resultType); err != nil {
+			return err
 		}
 
-		onMatches := func(in []protocol.CommitMatch) {
-			res := make([]result.Match, 0, len(in))
-			for _, protocolMatch := range in {
-				res = append(res, protocolMatchToCommitMatch(repoRev.Repo, j.Diff, protocolMatch))
+		g, ctx := errgroup.WithContext(ctx)
+		for _, repoRev := range page.RepoRevs {
```
Comment (tsenart) [inline internal/search/commit/commit_new.go:51]: @camdencheek: There's probably a better way to structure this loop to avoid the nested for loop. This was just the first pass.

File: internal/search/repos/repos.go [hunk 7]
```diff
@@ -39,17 +40,36 @@ type Resolved struct {
...
+	Opts                search.RepoOptions
 	DB                  database.DB
 	SearchableReposFunc searchableReposFunc
 }
 
+func (r *Resolver) Page(ctx context.Context, cursors []*database.Cursor) (Resolved, error) {
+	opts := r.Opts
+	opts.Cursors = cursors
+	if opts.Limit == 0 {
+		opts.Limit = 500
```
Comment (tsenart) [inline internal/search/repos/repos.go:68]: @rvantonder, @camdencheek: Thinking to make the default page size 500 to start with.

File: internal/search/repos/repos.go [hunk 8]
```diff
@@ -159,15 +180,24 @@ func (r *Resolver) Resolve(ctx context.Context, op search.RepoOptions) (Resolved
...
-		go func() {
-			excludedC <- computeExcludedRepositories(ctx, r.DB, op.Query, options)
-		}()
+		excludedC := make(chan ExcludedRepos, 1)
+		// Only compute excluded repositories if we don't have a limit.
+		if op.Limit == 0 {
+			// PERF: We Query concurrently since Count and List call can be slow
+			// on Sourcegraph.com (100ms+).
+			go func() {
+				excludedC <- computeExcludedRepositories(ctx, r.DB, op.Query, options)
```
Comment (tsenart) [inline internal/search/repos/repos.go:196]: @rvantonder: This excluded repos query doesn't really make sense when doing pagination. I'd actually like to move this out to its own method, which can then get called in `doResults` once. However, this count is actually quite slow at scale, so it'd be great to think of product solutions that would remove the need for it.

File: internal/search/repos/repos.go [hunk 9]
```diff
@@ -179,6 +209,29 @@ func (r *Resolver) Resolve(ctx context.Context, op search.RepoOptions) (Resolved
 			return Resolved{}, err
 		}
 	}
+
+	var next []*database.Cursor
```
Comment (tsenart) [inline internal/search/repos/repos.go:213]: This block is where we compute the next page cursors. The reason we need a slice is because star counts aren't unique, so we need a second column to paginate through.

File: internal/search/unindexed/structural.go [hunk 10]
```diff
@@ -157,23 +159,41 @@ type StructuralSearch struct {
...
 
-func (s *StructuralSearch) Run(ctx context.Context, stream streaming.Sender, repos []*search.RepositoryRevisions) error {
-	request, ok, err := zoektutil.OnlyUnindexed(repos, s.ZoektArgs.Zoekt, s.UseIndex, s.ContainsRefGlobs, s.OnMissingRepoRevs)
-	if err != nil {
-		return err
-	}
-	if !ok {
-		request, err = zoektutil.NewIndexedSubsetSearchRequest(ctx, repos, s.UseIndex, s.ZoektArgs, s.OnMissingRepoRevs)
+func (s *StructuralSearch) Run(ctx context.Context, stream streaming.Sender, repos searchrepos.Pager) error {
+	var cursors []*database.Cursor
```
Comment (tsenart) [inline internal/search/unindexed/structural.go:163]: @rvantonder: I noticed how much code here is identical to `TextSearch.Run`. I think there's an opportunity for reuse, but I didn't pursue it so far.

File: internal/types/types.go [hunk 11]
```diff
@@ -389,16 +389,18 @@ func (rs Repos) Filter(pred func(*Repo) bool) (fs Repos) {
...
 }
 
-// RepoName represents a source code repository name and its ID.
+// RepoName represents a source code repository name, its ID and number of stars.
 type RepoName struct {
-	ID   api.RepoID
-	Name api.RepoName
+	ID    api.RepoID
+	Name  api.RepoName
+	Stars int
```
Comment (tsenart) [inline internal/types/types.go:396]: We need this additional Stars column because we need to know the cursor value of the last record in a page, which can include star counts.

File: cmd/frontend/graphqlbackend/search_suggestions.go (hunk 5 above)
Comment (novoselrok) [reply cmd/frontend/graphqlbackend/search_suggestions.go:253]: That's right, the client code is not using the GQL suggestions endpoint anymore. We would like to remove the backend code as well, although I don't think we have a timeline for it yet.

File: cmd/frontend/graphqlbackend/search_alert.go (hunk 2 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search_alert.go:305]: Two observations: (1) this code is old and not really great. I'm not strongly attached to what we do here, we can see what happens. (2) We've recently had very strong feedback that fuzzy numbers of "repositories searched" or number of search results is not good, it needs to be exact. It's difficult to say how much (2) matters in the context of this alert, but it might be really important to some users, despite how expensive it is. 

Basically, the conservative approach is: keep this expensive++++ thing, or roll the dice until we get complaints if we remove it from alerts. I don't feel strongly, and I don't think anyone else on the team does either.

File: cmd/frontend/graphqlbackend/search_results.go (hunk 3 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search_results.go:1672]: Yes, please leave for me to migrate soon.

File: internal/search/repos/repos.go (hunk 7 above)
Comment (rvantonder) [reply internal/search/repos/repos.go:68]: sounds good :-)

File: internal/search/repos/repos.go (hunk 8 above)
Comment (rvantonder) [reply internal/search/repos/repos.go:196]: Moving it makes sense. 

> However, this count is actually quite slow at scale, so it'd be great to think of product solutions that would remove the need for it.

I think there's high potential that this tilts some users if we remove looking up excluded repos, based on strong feedback that the fuzzy "x results in y repositories" in our web app, where `y` is fuzzy, is not good. Excluded repos falls under a very similar banner, users _sometimes_ really really really want to know if results are incomplete, I don't know that we can just punt on this. They probably need it for auditing whether something is removed entirely, or finding how many uses there are of something, or finding all instances to change code or something. And if we don't reveal all the exact information of what is missed, they get (understandably) upset for this kind of use.

For the time being I (as in, my personal opinion, on that of others) would probably not roll the dice on this one and try preserve it while we figure out how we can gracefully handle reporting missed results. My guess is that it's only useful to to do the expensive thing in some cases, but for typical search a lot of users probably don't care, and we should give users a way to do that when they really do want it.

File: internal/search/unindexed/structural.go (hunk 10 above)
Comment (rvantonder) [reply internal/search/unindexed/structural.go:163]: Yeah, it's a bit duplicated but we're not ready to "crunch crunch" yet, it will happen later. It's because I'm migrating one search backend at a time, it's just not worth deduping right now. I think it's quite OK if we have a small number (`< 7`) duplicated code blocks for each job, and there are small differences that make it easier to reason about when it's separate.

File: cmd/frontend/graphqlbackend/search.go (hunk 1 above)
Comment (tsenart) [reply cmd/frontend/graphqlbackend/search.go:282]: Thanks for the insight! Two thoughts:

1. Perhaps we will find that caching is no longer needed if each page of repos takes a comparably tiny amount of time to execute compared to resolving ALL repos.
2. If we still want to do caching, I think we can be much smarter about it than just placing resolved repos in a field of the `searchResolver`. I would move the caching to the `searchrepos.Resolver` and use a subset of `RepoOptions` as the cache key, in which case we no longer need the current cache invalidation logic. Perhaps we could even use the full sub-expression query string as the cache key + the cursor. Thoughts?

File: internal/search/repos/repos.go (hunk 8 above)
Comment (tsenart) [reply internal/search/repos/repos.go:196]: OK, let's keep it but move to a separate method.

File: cmd/frontend/graphqlbackend/search.go [hunk 12]
```diff
@@ -330,6 +334,8 @@ func (r *searchResolver) suggestFilePaths(ctx context.Context, limit int) ([]Sea
 		return nil, nil
 	}
 
+	// TODO(tsenart): Move this to use a searchrepos.Pager directly and to not do any file path
+	//  suggestions if more than 1 repo filters are defined.
```
[the remark this answers is not in the document]
Comment (tsenart) [reply cmd/frontend/graphqlbackend/search.go:338]: @novoselrok: I guess this code will also go away? Can we make a plan to remove GQL suggestions? Glad to help.

[the remark this answers is not in the document]
Comment (novoselrok) [reply cmd/frontend/graphqlbackend/search.go:338]: Yep, seems like it can be removed too.

We should probably deprecate the `suggestions` resolver in the current release and then start removing it the next release. Does search-core want to handle the removal?

File: internal/search/commit/commit_new.go (hunk 6 above)
Comment (camdencheek) [reply internal/search/commit/commit_new.go:51]: Looks good to me for now 👍

File: cmd/frontend/graphqlbackend/search.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tsenart) [reply cmd/frontend/graphqlbackend/search.go:338]: Happy to remove some code! How would we go about the deprecation?

File: cmd/frontend/graphqlbackend/search.go (hunk 1 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go:282]: Yeah, both thoughts make sense to me. Seems like we would pursue (2) based on (1). 

> Perhaps we could even use the full sub-expression query string as the cache key + the cursor.

Yeah, could do. Strictly speaking we wouldn't need to use the full sub-expression, just the `repo`-relevant filters for each subexpression, which map to `RepoOptions`. Though, I would stick to "don't keep representation of query parse tree / expressions around when we're ready to run things", so I would probably favor caching on `RepoOptions` that represents our internal repo-backend-query datatype.
